File: app/lib/noise/base64url.dart
import 'dart:convert';
import 'dart:typed_data';

/// RFC 4648 base64url（无填充）。与桌面 `Base64Url.cs` 一致：
/// 二维码 / JSON 里承载二进制公钥、配对码用。
class Base64Url {
  static String encode(List<int> data) =>
      base64Url.encode(data).replaceAll('=', '');

  static Uint8List decode(String s) {
    final pad = (4 - s.length % 4) % 4;
    return base64Url.decode(s + ('=' * pad));
  }
}

File: desktop/CredentialTyper.Transport/Wire/Base64Url.cs
namespace CredentialTyper.Transport.Wire;

/// <summary>RFC 4648 base64url（无填充）。二维码 / JSON 里承载二进制公钥、配对码用。</summary>
public static class Base64Url
{
    public static string Encode(ReadOnlySpan<byte> data) =>
        Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    public static byte[] Decode(string s)
    {
        string b64 = s.Replace('-', '+').Replace('_', '/');
        switch (b64.Length % 4)
        {
            case 2: b64 += "=="; break;
            case 3: b64 += "="; break;
        }
        return Convert.FromBase64String(b64);
    }
}
